Improve CIDR allocation change resolution

tgraf opened the pull request:
When the allocation CIDR changes, the existing code errored out. When that
happened on the first agent restart, all existing endpoints failed to restore
due to failure to allocate the existing endpoint's IP address. The failure to
restore would then in turn get all existing endpoints deleted.

This is bad in two ways:

1. It causes all endpoints to be removed, causing the pods to restart
2. It does not allow the agent to ever come up again without intervention from
the user.

This commit solves problem #2 to allow nodes to recover automatically. It does
not solve problem #1 yet and requires pods to restart.

Solving problem #2 only makes sense once we have isolated all datapath aspects
such as the DNS proxy and L7 proxies out of the agent as endpoint connetivity
cannot be guaranteed if the agent cannot come up within a reasonable amount of
time.

Fixes: #6215

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6874)
<!-- Reviewable:end -->

aanm commented on daemon/daemon.go:
```diff
@@ -916,6 +917,59 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
+
+		// Remove the old host device and
+		var link netlink.Link
+		link, err = netlink.LinkByName(defaults.HostDevice)
```
replace with `option.Config.HostDevice`

tgraf replied:
@aanm Are you sure that you want to delete the flannel device?

brb replied:
Can we introduce the flag `doNotDelOldHostDevice` (set in the flannel mode) to the method, and then replace as suggested ^^?

tgraf replied:
I've added the check. We can work on a `doNotDelOldHostDevice` if we have more than one use case. I'm not sure about IPVLAN yet.

aanm commented on daemon/daemon.go:
```diff
@@ -916,6 +917,59 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
 	return counter
 }
 
+func (d *Daemon) prepareAllocationCIDR(family datapath.NodeAddressingFamily) (routerIP net.IP, err error) {
```
Please avoid running this if agent is being executed with flannel mode.

tgraf replied:
Sure, can do but it was run before in flannel mode.

aanm reviewed (dismissed):
I think this is a good approach. 2 main concerns, if the cidr changes between restarts:

- Old endpoints will immediately lose connectivity because their gateway disappeared and we should always avoid that, that was the main motivation behind preventing the cilium-agent from starting.
- We can ignore the previous point *only* if 0 pods are running at that time in the node.

tgraf committed:
ipam: Treat inability to allocate internal IP as fatal error

When the node internal IP is part of the allocation range, it must be reserved
in the allocation range and may not be used as an endpoint IP. When the
allocation to reserve the IP fails, the agent must quit as it risks to to
allocate it for an endpoint and thus creating an IP conflict.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Automatically change the allocation CIDR

When the allocation CIDR changes, the existing code errored out. When that
happened on the first agent restart, all existing endpoints failed to restore
due to failure to allocate the existing endpoint's IP address. The failure to
restore would then in turn get all existing endpoints deleted.

This is bad in two ways:

1. It causes all endpoints to be removed, causing the pods to restart
2. It does not allow the agent to ever come up again without intervention from
   the user.

This commit solves problem #2 to allow nodes to recover automatically. It does
not solve problem #1 yet and requires pods to restart.

Solving problem #2 only makes sense once we have isolated all datapath aspects
such as the DNS proxy and L7 proxies out of the agent as endpoint connetivity
cannot be guaranteed if the agent cannot come up within a reasonable amount of
time.

Fixes: #6215

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Error out if IPv4 allocation CIDR is not configured and IPv4 is enabled

It's always automatically generated when IPv4 is enabled but fail if it is not
set in case the automatic generation is ever disabled.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
option: Provide defaults for EnableIPv4 and EnableIPv6 in DaemonConfig

This is required to rely on option.Config.Enable{4,6} in daemon unit tests

Signed-off-by: Thomas Graf <thomas@cilium.io>
